### DISCUSSION TITLE
`juice_set_local_description` usage order clarification

### DISCUSSION THREAD

**Discussion by tmathews** [paullouisageneau/libjuice#281] [Q&A] [answered]
Per discussion here https://github.com/paullouisageneau/libdatachannel/issues/536#issuecomment-1000354916, it sounds like the order of usage for `juice_set_local_description` and `juice_set_remote_description` matters to resolve `ICE Role conflict (both controlling)` errors. 

I currently run into this warning, yet still get successful connections established. I would like clarity if this order is required to be correct even upon successful connections. 

Many thanks for any advice. Cheers.

**Comment by paullouisageneau** [comment, marked as answer]
The warning means both peers think they are the initiator of the connection. It can be recovered at the price of an added round-trip, but it indicates a more fundamental issue in the way signaling is done.

ICE is designed around an initiating agent and responding agent. libjuice handles it automatically: the initiator calls `juice_get_local_description()` to send its description to the responder, the responder first sets the initiator's description with `juice_set_remote_description()`, then calls `juice_get_local_description()` to send its description back to the initiator which sets it with `juice_set_remote_description()`. 

This ensures proper parameter negotiation and role selection. The collision if both agents initiate the connection at the same time is handled as an ICE conflict.

**Comment by tmathews** [comment]
Thank you for the detailed answer.
